File: 0373_kSmallestPairs.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        heap = []
        def insert(i, j):
            if i < len(nums1) and j < len(nums2):
                heapq.heappush(heap, (nums1[i] + nums2[j], i, j))
        insert(0, 0)
        kPairs = []
        while k and heap:
            summ, x, y = heapq.heappop(heap)
            kPairs.append([nums1[x], nums2[y]])
            insert(x, y + 1)
            if y == 0: insert(x + 1, y)
            k -= 1
        return kPairs

# Alternative solution (which gives MLE error)

class Solution1:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        heap = []
        for i in nums1:
            for j in nums2:
                heapq.heappush(heap, (i + j, i, j))
        result = []
        for _ in range(k):
            if heap:
                summ, x, y = heapq.heappop(heap)
                result.append([x, y])
            else:
                break
        return result
        
# Alternative solution (which gives TLE error)

class Solution2:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        return heapq.nsmallest(k, ((a, b) for a in nums1 for b in nums2), key = sum)

File: 0001-0500/0373_kSmallestPairs.java
class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        int m = nums1.length, n = nums2.length;
        List<List<Integer>> output = new ArrayList<>();
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        
        for (int i = 0; i < n; i++) {
            heap.offer(new int[] {nums1[0] + nums2[i], 0, i});
        }

        while (true) {
            int[] pair = heap.poll();
            k--;
            output.add(new ArrayList<>() {{
                add(nums1[pair[1]]); 
                add(nums2[pair[2]]);
            }});

            if (k == 0) {
                return output;
            }

            if (pair[1] + 1 < m) {
                heap.offer(new int[] {nums1[pair[1] + 1] + nums2[pair[2]], pair[1] + 1, pair[2]});
            }
        }
    }
}
